### ISSUE TITLE
Tests grouping

### ISSUE THREAD

**Issue by Guiguiprim** [nextest-rs/nextest#101]
My next maybe weird feature request.

This is somehow related to https://github.com/nextest-rs/nextest/issues/27 or at least could be a first step for it.

## My initial use case

I have a bunch of tests (across multiple binaries) who need to uniquely acquire a system resource. Currently I run the tests with `-j1` which is sub-optimal. I could run my tests in two passes: `cargo nextest -j1 <some filtering>` and `cargo nextest <opposite filtering>`, but it's not ideal too and force to maintain two filtering (hopefully without missing some tests).

Having a way of defining test groups with custom execution configuration would be great.

## Proposed feature

Adding groups support !

It could allow:
- only running predefined groups of tests
- running differently different groups of tests
- more ?

### Defining groups

I think it could go inside `./config/nextest.conf`. Maybe like:

```rust
struct CustomProfileImpl {
  // ...
  groups: Vec<TestGroupConfig>,
}

struct TestGroupConfig {
  name: String,
  include: Vec<String>, // filters
  exclude: Vec<String>, // filters
  test_threads: Option<usize>, // what I need
  processes: ExecutionProcesses, // for #27 
  // maybe more option:
  // - execution config (group run concurrently with other group or not, ...)
}

enum ExecutionProcesses {
  OneByTest,
  // execute all test from the same binary in the same process
  // two binaries can be run at the same time
  OneByBinaryParallel,
  // execute all test from the same binary in the same process
  // two binaries can NOT be run at the same time
  OneByBinarySequentiel,
}
```

### Constructing groups

Constructing groups could be as easy as:

- Applying `TestBuildFilter` to get the tests pool
- For every `TestGroupConfig` construct a group by taking out of the pool the matching tests
- All the remaining tests are put in a unnamed group with the config give to the CLI

This would give an importance to the group definitions order while also avoiding any duplication.

### Running tests

Here things can be as complex as we want.
- In the first version I would go with running groups one after the other.
- In following version we could add more complex scheduling scheme
    - If we have 3 groups with `-j1` maybe they are allowed to run concurrently
    - ...

### New CLI options

With all of this, we could probably add some options (for both `list` and `run`): `--group <NAME>` and `--unamed-group` (or any better name) to only run tests within some groups.

**Comment by Guiguiprim** [comment]
Again I might be able get the v1 of this feature a go.

**Comment by poliorcetics** [comment]
Test grow organically and are often moved and renamed. I don't know if there is an easy way to say "this test belong to this group" and "this one does not belong to any group, i.e the default".

Can proc-macros access the full path of a function ? If yes, it becomes possible to do it with something like `#[nextest(group = "my-non-default-group")]` on tests we want marked

**Comment by Guiguiprim** [comment]
As of today, `cargo-nextest` doesn't need to instrument the code in any way. And I'm not even sure what this attribute would do and how it would be visible from within `nextest-runner`. The only thing it could do is changing the test name I guess.

In my case, I think I could have a convention of having `_use_rsc_name_` in the test name and filter based on this. Maybe I could go the extra miles by creating myself a attribute for this (which could rename the test and wrap the implementation and give it an handle to my resource, to make using it the nice way to go).

**Comment by poliorcetics** [comment]
Yes, naming convention needs the least engineering and would not depend on `cargo`'s internals, which is better in the long term.

For quick CLI grouping, filtering would stay available and groups would only be accessible through the config file, is that right ?

**Comment by sunshowers** [comment]
Thanks for the request!

Some thoughts about this:

**Group scope.** So there's three potential choices of group units:
1. Individual tests.
2. Individual test binaries.
3. Workspace packages (aka crates).

I think 3 (workspace packages) is correct here for the reasons outlined below. I think it's reasonable to say that tests which need special handling should be moved into their own crate(s).

**Semantics.** Keeping it restricted to 3 simplifies many issues like feature unification. 1 and 2 both make it harder to define semantics here. This also minimizes CLI surface, since we can piggyback on the `--package` option that already exists (see the CLI surface section below).

**Configuration.** I think this should be done via a separate config file, e.g. `.config/cargo-query.toml` (not exactly sure about this name). Why? Because groups of workspace packages is more generally useful than just nextest -- you could want to `cargo build` or `cargo check` specific groups as well. We shouldn't foreclose those use cases.

The config file should look something like:

```toml
[groups]
my-group = ["package1", "package2"]
group2 = ["package3", "package4"]
# See the CLI surface for more about this.
group-union = "group(my-group) + group(group2)"
```

**CLI surface.** I really think this should piggyback on `-p` and be done as a [DSL](https://en.wikipedia.org/wiki/Domain-specific_language). The reason is that a DSL offers a ton of orthogonality (now and in the future) without polluting the argument namespace. So for example:
* If you want to run all the tests in a group, you say: `cargo nextest run -p 'group(my-group)'`.
* If you want to run all the tests in the transitive dependencies of a group, you say e.g. `cargo nextest run -p 'deps(group(my-group))'`.
* If you want to run all the tests in the transitive dependencies of a *package*, you use `'deps(package1)'`. (This is what I mean by orthogonality -- the logic done for groups also automatically applies for packages.)

One thing to exploit here is that package names cannot now, or in the future, contain whitespace and syntax characters other than `-` (see the [Rust documentation on non-ascii-idents](https://rust-lang.github.io/rfcs/2457-non-ascii-idents.html#conformance-statement)). This means that we can use those whitespace and syntax characters in the DSL for our own purposes.

Some DSL prior art that I think is great:
* [bazel query](https://docs.bazel.build/versions/main/query.html)
* [buck query](https://buck.build/command/query.html)
* [Mercurial revsets](https://www.mercurial-scm.org/repo/hg/help/revsets) -- note particularly the boolean operators, aliases and shortcuts here.

**Running tests.** So this I think is really a separate question and should be tackled separately in #27. In particular:
* the unit for *this* is a test binary and not a crate
* this is specific to nextest and not more generally useful.

Note that once you have groups, you can just do `cargo nextest -j1 -p 'group(my-group)'` and `cargo nextest -p 'not group(my-group)'`, which neatly solves the immediate problem.

I realize this is a lot of work, but nextest is going to have to maintain this UI for a long time so I think it's important to get this right.

**Comment by Guiguiprim** [comment]
Saddly my workspace contains one big crate that still need to be splitting up, work is ongoing but low priority. But this is a me problem not a you problem. 😉

I think the idea of a custom attribute to add a prefix to the test named and give a mandatory resource token to help with filtering as discussed with @poliorcetics would work nicely for me.

One point that I had, was have only one nextest call (with `--no-fail-fast`) makes it possible to have all tests results at once. Splitting it across multiple calls would fail the CI without running all tests. But some CI wizardy could probably fix this.

Otherwise I do like your design and it might indeed be the right way to go for nextest.
